### DISCUSSION TITLE
are hotkeys userdefinable / can I assign a hotkey to any function?

### DISCUSSION THREAD

**Discussion by StefanL38** [xournalpp/xournalpp#2778] [Ideas]
Hi everybody,

not really sure if this is the right place to ask this.
So if not post a link to the place where to ask such questions.

I would like to use Xournal++ on a lenovo X1-tablet with digital pen
of course this basically works.
Now to increase productivity I want to use a smartphone-app like inifiniMote that sends keystrokes or even key-sequences to my tablet. My writehand stays at writing position and my non-write-hand taps on the smartphone-screen to do things like
add new page and paste clibboard-content by a single tap. Changing color or pen-function  etc. etc. each by a single tap. 

So my questionis:  are the hotkeys user-definable?   ANd if yes can I assign a hotkey to any function?
especially I want to be able to change the pen-color with hotkeys. 

Maybe there is a plugin that can do so?
If you use two-stroke-hotkeys a lot of functions can be assigned.

best regards Stefan

**Comment by rolandlo** [comment]
Yes, you can write a plugin where you assign hotkeys (accelerators) to all functions you named (like adding a new page, pasting the clipboard-content, changing color or pen-function). Have a look at the [documentation of plugins](https://xournalpp.github.io/guide/plugins/plugins/) on the work-in-progress website. Feel free to ask any further questions.

**Reply by StefanL38** [reply]
Hi Rolando,
thank you very much for pointing me to the plugin website.
I have quite some experience in programming Delphi, and C++ (for Arduino) SO I know the basic concepts of programming.
Though this LUA seems to be quite different with tables and lists. Do there exist any kind of plugins yet?
Where could I ask for a working example of a plugin to see how things work together between plugin-folder, ini-file lua.main-file etc. 

Learning LUA from scratch would be probably a very long journey to finally be able to write Xournal++-plugins
I estimate if I can read a well commented example on how to use the commands and structure I will be up and running within 10 to 20 hours of reading and experimenting

Maybe even a plugin for another software that uses the same basic tecniques will be useful.

best regards Stefan

**Comment by rolandlo** [comment]
It's actually pretty easy, if you just need some shortcuts for things that you can already do from Xournal++'s GUI. See some examples in [the plugins folder](https://github.com/xournalpp/xournalpp/tree/master/plugins). Another example (that may be particular useful for you) is in [this issue](https://github.com/xournalpp/xournalpp/issues/919#issuecomment-700247463)

Just to make clear: you need version 1.1.0+dev (see [nightly releases](https://github.com/xournalpp/xournalpp/releases/tag/nightly)) for most of this to work.

**Reply by StefanL38** [reply]
Hi Rolando,
thank you very much for pointing me to the right version. I took a look into the cycle-color-plugin.
I got a first grasp of how it works. 
The documentation is pretty short.  Is there any further documentation of the parameters?

Does there already exist a place on the internet where plugins can be discussed and questions been asked about how to program plugins?

If such a place doesn't exist yet; what is your suggestion where to create such a "meeting-point" ?
best regards Stefan

**Comment by rolandlo** [comment]
The meeting point is right here. :-) Don't hesitate to ask questions.
There is some further documentation of the usage of the plugin API in the code file [luapi_application.h](https://github.com/xournalpp/xournalpp/blob/master/src/plugin/luapi_application.h) (which is linked on the website as well). Every API command has at least one or two examples which demonstrate how to use the parameters.

**Reply by StefanL38** [reply]
The Test123-example-plugin has some commands print
`-- This is an example Xournal++ Plugin - copy this to get started

var_dump = require "var_dump"

-- Register all Toolbar actions and intialize all UI stuff
function initUi()
  print("Hello from Example: Plugin initUi called\n");

  ref = app.registerUi({["menu"] = "Test123", ["callback"] = "exampleCallback", ["accelerator"] = "<Control>t"});
  print("Menu reference:");
  var_dump(ref);

  print("Example plugin registered\n");
end

-- Callback if the menu item is executed
function exampleCallback()
  result = app.msgbox("Test123", {[1] = "Yes", [2] = "No"});
  print("result = " .. result)
end
`
Where does it print to?  Is there some kind of console-window? If yes how can I make it visible
or is it just for debuging while running the code in Lua-for-windows?

What I'm missing is explaining text around the pure code.
I mean something like:
"let's imagine you want xournal++ to popup a window on a hotkey to let the user choose "yes" or "No" to a question.

The details described above In this example are still easy to understand. 
This example shall show what kind of explaining text I mean.

But there are some details that I don't understand without a generalised description
example that I'm unsure about
`  ref = app.registerUi({["menu"] = "Test123", ["callback"] = "exampleCallback", ["**accelerator**"] = "<Control>t"});
`
is "accelerator" the keyword for "hotkey" ?? or something else.
sure a hotkey accelerates execution but accelerator is not a common synonym for hotkey?

I try to translate the given example to a generalised description
  ref = app.registerUi({["menu"] = "Test123", ["callback"] = "exampleCallback", ["accelerator"] = "<Control>t"});

app.registerUi(({["menu"]

"menu": type of GUI-element (what are the other possible types of UI-elements? where can I look up them for reference?

"Test123"    Text shown as the menu-item when clicking on main-menu Plugin

["callback"]  :   defining type of ... what? ---
"exampleCallback"    name of the function executed if ....?

understanding the underlying generalised structure needs more than a singleexample.
You can't abstract from a single example. A single example does not allow to distinguish between fixed terms and variable terms.

Another example that goes beyond that first example.
`result = app.msgbox("Test123", {[1] = "Yes", [2] = "No"});
if result == YES`  
  --examples of an additional action

From such a small example writing plugins will be reduced to Lua-experts. Anybody like me with experience in other programming-languages are forced to learn Lua from scratch.
Expanding the examples like described above will speed up running up the learning-curve a lot for C++ or whatever other language  coders.
So expanding on well commented examples is a strategic investion into a growing user-community.
I'm willing on contributing to that. Let's see how much outstanding features we can add to xournal++ over microsoft whiteboard etc.
best regards Stefan

**Reply by rolandlo** [reply]
> Where does it print to? Is there some kind of console-window? If yes how can I make it visible
> or is it just for debuging while running the code in Lua-for-windows?
> 
On Linux it prints to the terminal console, if you run `xournalpp` from the terminal. On Windows you can only see these messages if you build xournalpp from source and run `xournalpp` from the MinGW console. It's the same thing with all the `g_messages` in the C++ code. 

> ` ref = app.registerUi({["menu"] = "Test123", ["callback"] = "exampleCallback", ["**accelerator**"] = "<Control>t"});`
> is "accelerator" the keyword for "hotkey" ?? or something else.
> sure a hotkey accelerates execution but accelerator is not a common synonym for hotkey?
> 
This is Gtk terminology, which distinguishes between "accelerators" and "mnemonics", see [Gtk reference](https://developer.gnome.org/gtk3/stable/gtk3-Keyboard-Accelerators.html). Basically it is what you would call hotkey.

> I try to translate the given example to a generalised description
> ref = app.registerUi({["menu"] = "Test123", ["callback"] = "exampleCallback", ["accelerator"] = "t"});
> 
> app.registerUi(({["menu"]
> 
> "menu": type of GUI-element (what are the other possible types of UI-elements? where can I look up them for reference?
> 
> "Test123" Text shown as the menu-item when clicking on main-menu Plugin
> 
> ["callback"] : defining type of ... what? ---
> "exampleCallback" name of the function executed if ....?
> 
> understanding the underlying generalised structure needs more than a singleexample.
> You can't abstract from a single example. A single example does not allow to distinguish between fixed terms and variable terms.
> 
It's explained on the [plugin section](https://xournalpp.github.io/guide/plugins/plugins/#plugin-structure). Here `"menu"`, `"callback"`, and `"accelerator"` are the fixed terms. Those are the keys of a Lua table that are passed to the `registerUi` function defined in the plugin API. There are no other keys for this function. "Test123", "exampleCallback" and "t" are variable terms. In this example you should have a 
```Lua
function exampleCallback()
  -- SOME CODE GOES HERE
end
```
defining the code to be executed when either the menu entry for "Test123" in the Plugin menu is clicked or if the accelerator (hotkey) "t" is pressed.

Note by the way that "t" is not a very good hotkey, since it will create problems using the Text tool in Xournal++.

> Another example that goes beyond that first example.
> `result = app.msgbox("Test123", {[1] = "Yes", [2] = "No"}); if result == YES`
> --examples of an additional action

Here a message box will be displayed, showing the Text "Test123" and buttons "Yes" and "No". If you press the button "Yes", then the `result` variable will be set to 1, if you press "No", it will be set to 2.
 
> From such a small example writing plugins will be reduced to Lua-experts. Anybody like me with experience in other programming-languages are forced to learn Lua from scratch.

I can assure you that I'm not a Lua expert either and I had no previous experience with Lua before starting to work on plugins and to extend the plugin API. There is little you need to know about Lua. The most essential thing is to understand Lua tables. The problems you encountered has more to do with the plugin API we provide than with Lua as a language.

> Expanding the examples like described above will speed up running up the learning-curve a lot for C++ or whatever other language coders.
> So expanding on well commented examples is a strategic investion into a growing user-community.
> I'm willing on contributing to that. 

You are welcome to do that. Sure there is a lot of space for improvement. Just gain some experience with it.

Let's see how much outstanding features we can add to xournal++ over microsoft whiteboard etc.
> best regards Stefan

**Reply by StefanL38** [reply]
Hi Roland,

thank you very much for answering. 

For activating a new plugin I have to restart xournal++.  Even for a small change in the code I have to restart xournal++. 
Is there a way to speed up this? I don't know what is done "under the hood" restarting is a clean way to do it. (and by the way re-booting / re-starting is the most common task all modern electronik equipment and software is doing) but if there is a chance to have a feature-"development mode" where it is enough to press a - guess what - hotkey to reload the plugin-code - this would be great.

some more questions:
am I right the lua-code of the plugin must have the name "main.lua"  ?  
other names don't work?

are there any conventions for the sub-foldername similar to libraries in Arduino-C++?

if the plugin has syntactical errors the plugin just does not apear
How about showing an error-message what the error is?
This error-message should be shown **inside** xournal++. It might be that there is something in the console. I don't know because I'm a windows user. 

quote (by the way how is real quoting done in this editor-field there is no button for it?)
_"On Windows you can only see these messages if you build xournalpp from source and run xournalpp from the MinGW "_
**This is something I just won't do. It sounds like another 50 hours of learning to get it up and running**

Is there a possability to have **two**-key-strokes? I'm used to use two-key-stroke-shortcuts in CAD-software  "li"  draw line    "sg"   selection grid,   "se" selection endpoints etc. Having two-key-shortcuts enables adding a lot of functionality.

And I would like to add a user-configurable "hot-button-bar"  a small area (size user-adjustable, number of buttons user-adjustable, the buttons can have user-definable smallgrafics/Icons) where you can define actions like "copy" , "highlight with green"  or whatever the user wants to have one single left-hand-tap away  while her/his right-hand with pen stays right at that position of the screen she/he wants to write on.

I guess the hot-button-bar is a bit more than showing a message-box. But that would out-perform any other pen-based application with increased productivity. 

I think I have a lot of useful ideas:
a plugin that can activate blinking a layer ON/OFF to focus the attention of viewers to a certain point through making visible / unvisible a highlightet area on the screen or an additional textbox or whatever is on the layer. In sequence this would even enable multiple-step animations. Of course there is a lot of animation software but they are highly complex = oversized for a three-step animation.

As the user-interface is lua-based can you recommend a Lua-tutorial which focuses on things like "show new window"  with icon-buttons and how to code events if button is pressed?
best regards Stefan

**Reply by StefanL38** [reply]
Another question about the accelerator-item (aka hotkey)  
where can I find a list where alle keys like Control, Alt, Shift, Windows-key and how to use combinations of them are listed. I looked through the gtk accelerator-group. This website shows all details what is done "under the hood"  but this site does not explain how a key-combination like Windows-Alt-G   would be written in the Lua-code

best regards Stefan

**Reply by rolandlo** [reply]
> For activating a new plugin I have to restart xournal++. Even for a small change in the code I have to restart xournal++.
> Is there a way to speed up this? I don't know what is done "under the hood" restarting is a clean way to do it. (and by the way re-booting / re-starting is the most common task all modern electronik equipment and software is doing) but if there is a chance to have a feature-"development mode" where it is enough to press a - guess what - hotkey to reload the plugin-code - this would be great.
> 
I agree this would be great. Maybe not that hard to achieve. I will have to look into it, once I find time for it.

> some more questions:
> am I right the lua-code of the plugin must have the name "main.lua" ?
> other names don't work?
No you can give them a different name. It's just the name that you define in the `plugin.ini` file.
> 
> are there any conventions for the sub-foldername similar to libraries in Arduino-C++?
> 
No, there are no conventions thus far.

> if the plugin has syntactical errors the plugin just does not apear
> How about showing an error-message what the error is?
> This error-message should be shown **inside** xournal++. It might be that there is something in the console. I don't know because I'm a windows user.

Yes, there is something in the console, but not that useful either. This could be improved for sure.

> 
> quote (by the way how is real quoting done in this editor-field there is no button for it?)

Quoting works with a ">" sign and a space before the quoted text like usual in [MarkDown](https://en.wikipedia.org/wiki/Markdown).

> _"On Windows you can only see these messages if you build xournalpp from source and run xournalpp from the MinGW "_
> **This is something I just won't do. It sounds like another 50 hours of learning to get it up and running**
> 
It takes some time, I agree, but it is described step by step in this [wiki](https://github.com/xournalpp/xournalpp/blob/master/readme/WindowsBuild.md)

> Is there a possability to have **two**-key-strokes? I'm used to use two-key-stroke-shortcuts in CAD-software "li" draw line "sg" selection grid, "se" selection endpoints etc. Having two-key-shortcuts enables adding a lot of functionality.

Hmm, I don't know of a way to do so.
 
> As the user-interface is lua-based can you recommend a Lua-tutorial which focuses on things like "show new window" with icon-buttons and how to code events if button is pressed?
> best regards Stefan

This cannot be done in pure Lua. You need some module like the [`lgi`-module](https://github.com/pavouk/lgi) to create a GUI. Unfortunately there is not much documentation about it. The most useful I found was the [gtk-demo](https://github.com/pavouk/lgi/tree/master/samples/gtk-demo) they provide. You have to run the `main.lua` file and then you can select the example and see its source code. I have used the lgi-module for #2724.

**Reply by rolandlo** [reply]
> Another question about the accelerator-item (aka hotkey)
> where can I find a list where alle keys like Control, Alt, Shift, Windows-key and how to use combinations of them are listed. I looked through the gtk accelerator-group. This website shows all details what is done "under the hood" but this site does not explain how a key-combination like Windows-Alt-G would be written in the Lua-code

All I know about this are the description of `gtk_accelerator_parse` and the [source file](https://github.com/GNOME/gtk/blob/master/gtk/gtkaccelgroup.c)
Windows-Alt-G should be `<Alt><Super>G` I think.

**Comment by Technius** [comment]
Maybe it would be useful to create a "Keybindings Hack" plugin that provides a keybinding GUI through `lgi`. It's a "hack" because the proper solution would be #919.

**Reply by StefanL38** [reply]
Hi Technius,
thank you for answering- never heard from lgi before. Can you provide a link that explains what it is?

EDIT: I guess I did find what it is: https://github.com/pavouk/lgi

Of course there are positive exceptions from my main experience with GitHub. User Roland is such an exceptions.
Thank you very much for all your support.

Anyway I want to state very clearly what I think about GitHUB:
wild collection of nerds uploading their code with none or minimal documentation
"use it as is"

My answer to lgi is:  Let the hub wobble all you GitHubbers  stay nerdy with all your **NON**-documentation

If I post a question in the Autohotkey-user-forum the probability to get even a direct code-example is pretty high
So my route is autohotkey (AHK) and 

app.registerUi({["menu"] = "Test123", ["callback"] = "exampleCallback", ["**accelerator**"] = "<Control>t"});  

to get the hotkeys working.

best regards Stefan
